File: ADT/djikstra's algo/3606.py
class Solution:
    def validateCode(self, code: str) -> bool:
        if not code:
            return False
        for ch in code:
            if not (ch.isalnum() or ch == '_'):
                return False
        return True
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        mp = {
            "electronics": 0,
            "grocery": 1,
            "pharmacy": 2,
            "restaurant": 3,
        }
        temp = []
        for i in range(len(code)):
            if isActive[i] and businessLine[i] in mp and self.validateCode(code[i]):
                temp.append((mp[businessLine[i]], code[i]))    

        temp.sort()
        result = [code for _, code in temp]

        return result
        
        

File: ADT/djikstra's algo/3606.cpp
//Approach (Using map and sorting)
//T.C : O(n*l + nlogn) where n = code.size() and l = average length of codes
//S.C : O(n)
class Solution {
public:
    bool checkValidCode(string &code) { // O(l)
        if(code.empty())
            return false;

        for(char &ch : code) {
            if(!isalnum(ch) && ch != '_') {
                return false;
            }
        }

        return true;
    }

    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        
        unordered_map<string, int> mp = {
            {"electronics", 0},
            {"grocery", 1},
            {"pharmacy", 2},
            {"restaurant", 3}
        };

        vector<pair<int, string>> temp; //{businessLineIntegerValue, code}

        for (int i = 0; i < code.size(); i++) { //O(n * l)
            if (isActive[i] &&
                mp.count(businessLine[i]) &&
                checkValidCode(code[i])) {

                temp.push_back({mp[businessLine[i]], code[i]});
            }
        }

        sort(begin(temp), end(temp)); //O(n.logn)
        vector<string> result;

        for(auto &it : temp) {
            result.push_back(it.second);
        }

        return result;
    }
};
